### DISCUSSION TITLE
Export to Excel Document

### DISCUSSION THREAD

**Discussion by ridwan-aswan** [bosskmk/pluto_grid#121] [Q&A]
Can I export the Pluto grid to an Excel document?

**Comment by jeansebastienZ** [comment]
It's technically possible yes , internally i don't know but externally i use this package csv: ^4.0.3 . If need some code you can ask .

**Comment by bosskmk** [comment]
There is no function to export to Excel yet.
It will be added, but it is not in the plan yet.
As jeansebastienZ said, implementing it externally is a quick way.

**Reply by ridwan-aswan** [reply]
I really hope this will be realized soon, because our clients really need to export to Excel for data processing according to their needs.

thanks @bosskmk

**Comment by henry2man** [comment]
@jeansebastienZ Can you share some code with us? I'm very interested on this and I think that a properly implementation should be embedded into the package for a lot of reasons: 

- Enable/Disable the export features could be as simple as setting some flag.
- You can reuse things like "formatters" in number cells
- You can define if a column is/is not "exportable" and custom "serializers" (imagine a column with a complex renderer that should be transformed into something with an string representation
- You can "embed" export UI options into the grid (in the footer, using context menus...)
- And in an advanced implementation, you'll be able to export complex excel-like formats with reordered columns, UI-like column widths, custom color in rows and so on, probably using [excel](https://pub.dev/packages/excel) package

Having your feet on the ground, we should focus on having something minimal, some kind of Minimum Viable Product:

- Only `csv` export using [csv](https://pub.dev/packages/csv) package
- Configuration:
  - `exportable` flag
  - `csv` inherited parameters
    -  `fieldDelimiter`
    - `textDelimiter`
    - `textEndDelimiter`
    - `eol`
- A minimal UI with no localization, something like "[📄 CSV]"
- With all visible columns
- Using existing serializers

We need this in our project so I can contribute to the package with a PR.

@bosskmk What are your thoughts on this?

**Reply by bosskmk** [reply]
Contributions are always welcome.
However, in order for the contributed code to be distributed, the following conditions must be met.

1. No problem with existing operation
2. The test code should succeed. (This is what I think I should do.)

**Reply by henry2man** [reply]
Ok, give me some time to see if I can do a quick prototype and then we can open a PR draft to review it.

**Reply by henry2man** [reply]
@bosskmk  Please take a look at https://github.com/bosskmk/pluto_grid/pull/425

**Reply by bosskmk** [reply]
@henry2man 
Deployed in version 2.10.0.
It would be nice if the function to import as csv would be added later. I haven't been able to add features these days because I don't have time, so I'll try to add them.

Thank you very much for your contribution.

**Reply by henry2man** [reply]
Great! You're welcomed. 

I think we can keep this discussion open in order to develop more advanced improvements.

**Comment by henry2man** [comment]
@bosskmk Hi again! I have an equivalent implementation for exporting to pdf, too. It's generic, but it needs two new packages (`pdf` & companion package `printing`). 

Add more features is always cool but, also, we should take care of the size of the package. Not everyone will need to export and/or print. 

Do makes sense to have these in a separate and optional `pluto_grid_plugins` package? Or, on the contrary, would it be to apply over-engineering at this point in the project?

Nevertheless tell me what are your thoughts about how can I add PDF export here ✌️

**Reply by bosskmk** [reply]
To enable extensibility and many contributions, it would be better to separate it in the form of a plugin.

I will create a separate team and invite you. It would be good to proceed with plugin development here.

In the future, it would be good to develop server-side pagination, the top menu bar for convenience, and a calendar using PlutoGrid (currently, the built-in calendar is inconvenient) in the form of a plug-in.

How about? Would you like to join us by creating a team? Or you can suggest another way.

**Reply by bosskmk** [reply]
Also, it's a private repository for now, but I'd like to create a team to empower people like you.

**Reply by bosskmk** [reply]
or simply,
Another way is to isolate the deployment by adding a folder called plugins in the current repository. I think this would be good too.

How to receive and distribute PR as it is now.

**Reply by henry2man** [reply]
I've seen this approach (one repo, many packages) in other packages (in fact, `pdf` and `printing` are a valid example of that). I'm not sure how this affects to https://pub.dev publishing but I feel this alternative is more convenient and simple. 

Tell me what you think and then I'll contribute the pdf export utility methods 🚀

**Reply by bosskmk** [reply]
@henry2man 
I have created a temporary branch.
If you look at the link below, there is a packages folder and pluto_grid_export is there.
You can add packages in this form. In a separate, segregated manner of distribution.

https://github.com/bosskmk/pluto_grid/tree/feature/3.0.0/pluto_grid_export

**Reply by henry2man** [reply]
In order to make things simple for now, I'm going to integrate PDF export into the same package (I feel this will be used in conjunction most of the time.). Also I've found some common code so I'm going to refactor code a bit.

**Reply by henry2man** [reply]
@bosskmk Please check this Draft PR https://github.com/bosskmk/pluto_grid/pull/440. The main issue is I had to reference "`pluto_grid`" dependency in `pluto_grid_export` as a "`path`" dependency (If you want we can continue the conversation directly in the PR)
